src/components/utils/src/threads/thread_qt.cc

```diff
@@ -66,25 +65,50 @@ void* Thread::threadFunc(void* arg) {
 
   // Sets thread id in order to be able to check that id on thread joining
   thread->handle_ = QThread::currentThread();
-  thread->thread_created_ = true;
+  thread->thread_state_ = kThreadStateIdle;
   thread->state_cond_.Broadcast();
 
-  while (!thread->finalized_) {
+  // We have special variable for controlling iterations/exiting thread
+  // in order to separate decision logic (continue iterations or exit?)
+  // from controlling while cycle
+  bool continueIterations = true;
+
+  while (continueIterations) {
     SDL_DEBUG("Thread #" << QThread::currentThreadId() << " iteration");
-    while (!thread->do_iteration_) {
+    while (kThreadCommandNone == thread->thread_command_) {
       thread->run_cond_.Wait(thread->state_lock_);
     }
     SDL_DEBUG("Thread #" << QThread::currentThreadId() << " execute. "
                          << "stopped_ = " << thread->stopped_
                          << "; finalized_ = " << thread->finalized_);
-    if (!thread->stopped_ && !thread->finalized_) {
-      thread->isThreadRunning_ = true;
-      thread->state_lock_.Release();
-      thread->delegate_->threadMain();
-      thread->state_lock_.Acquire();
-      thread->isThreadRunning_ = false;
+
+    switch (thread->thread_command_) {
+      case kThreadCommandRun:
+        thread->thread_state_ = kThreadStateRunning;
+        pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
```

Comment:
@orhan-mehmedov WOuld be nice to extract this case logic to private functiion with description